File: Platformer/scripts/Water.jl
using JulGame.MainLoop 

mutable struct Water
    offset
    parent
    player
    
    function Water()
        this = new()

        this.parent = C_NULL
        this.player = C_NULL
        this.offset = C_NULL

        return this
    end
end

function Base.getproperty(this::Water, s::Symbol)
    if s == :initialize
        function()
            this.offset = JulGame.Math.Vector2f(this.parent.getTransform().position.x + MAIN.scene.camera.offset.x, 5.5)
            this.player = MAIN.scene.getEntityByName("Player")
        end
    elseif s == :update
        function(deltaTime)
            this.parent.getTransform().position = JulGame.Math.Vector2f(this.player.getTransform().position.x, 0) + this.offset
        end
    elseif s == :setParent 
        function(parent)
            this.parent = parent
        end
    elseif s == :onShutDown
        function ()
        end
    else
        try
            getfield(this, s)
        catch e
            println(e)
        end
    end
end

File: Platformer/ts/_generated/scripts/Water.ts
import { registerScript } from "julgame/src/engine/runtime/scriptRegistry";
import { vecAdd, vecSub, vecMul, vecDiv, vecNeg } from "julgame/src/engine/core/vectorOps";


    // using JulGame

    export class Water {
        main
        offset
        parent
        
        constructor() {
            

            this.parent = null
            this.offset = {x: 0, y: 0}

        }
    }

    function JulGame_initialize_Water(self: Water) {
        self.offset = {x: self.parent.transform.position.x + 9, y: 7.5}
    }

    function JulGame_update_Water(self: Water, deltaTime) {
        self.parent.transform.position = vecAdd({x: (globalThis as any).MAIN.scene.camera.position.x, y: 0}, self.offset)
    }

    function JulGame_on_shutdown_Water(self: Water) {
    }
registerScript("Water", {
    create: () => new Water(),
    initialize: JulGame_initialize_Water,
    update: JulGame_update_Water,
    onShutdown: JulGame_on_shutdown_Water,
});
